sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/processor/CompactBlockProcessor.kt

```diff
@@ -301,149 +370,396 @@ class CompactBlockProcessor internal constructor(
+                }
+                checkAllBalances()
+
+                when (rangeSyncProgress.resultState) {
+                    SyncingResult.UpdateBirthday -> {
+                        updateBirthdayHeight()
+                        SyncingResult.AllSuccess
+                    }
+                    SyncingResult.EnhanceSuccess -> {
+                        Twig.info { "Triggering transaction refresh now" }
+                        // Invalidate transaction data and return the common batch syncing success result to the caller
+                        checkTransactions(transactionStorage = repository)
+                        SyncingResult.AllSuccess
+                    }
+                    is SyncingResult.Failure -> {
+                        rangeSyncProgress.resultState
+                    } else -> {
+                        // First, check the time and refresh the prepare phase inputs, if needed
+                        val currentTimeMillis = System.currentTimeMillis()
+                        if (shouldRefreshPreparation(
+                                lastPreparationTime,
+                                currentTimeMillis,
+                                SYNCHRONIZATION_RESTART_TIMEOUT
+                            )
+                        ) {
+                            SyncingResult.RestartSynchronization
+                        } else {
+                            // Continue with processing
+                            SyncingResult.AllSuccess
+                        }
+                    }
+                }
+            }.takeWhile {
+                syncingResult = it
+                it == SyncingResult.AllSuccess
+            }.collect()
+
+            when (syncingResult) {
+                is SyncingResult.AllSuccess -> {
+                    // Continue with processing the rest of the ranges
+                }
+                is SyncingResult.RestartSynchronization -> {
+                    // Restarting the synchronization process
+                    return BlockProcessingResult.RestartSynchronization
+                } else -> {
+                    // An error came - remove persisted but not scanned blocks
+                    val lastScannedHeight = when (val result = getMaxScannedHeight(backend)) {
+                        is GetMaxScannedHeightResult.Success -> result.height
+                        else -> null
+                    }
+                    lastScannedHeight?.let {
+                        downloader.rewindToHeight(lastScannedHeight)
+                    }
+                    deleteAllBlockFiles(
+                        downloader = downloader,
+                        lastKnownHeight = lastScannedHeight
+                    )
+                    return (syncingResult as SyncingResult.Failure).toBlockProcessingResult()
+                }
+            }
+        }
         return BlockProcessingResult.Success
```

Comment:
My review of this function was not particularly thorough; I'm missing a lot of the needed context.

Reply:
Yeah, I understand. Thank you anyway!